File: scripts/Subsets II/Subsets II.py
class Solution(object):
    def subsetsWithDup(self, nums):
        def fn(nums):
            if len(nums)==0:
                return [[]]
            ele=nums[0]
            without=self.subsetsWithDup(nums[1:])
            wit=[]
            for i in without:
                l=[ele]+i
                if l not in without:
                    wit.append(l)
            return without+wit
        return fn(sorted(nums))

File: algorithms/Subsets II/Subsets II.java
class Solution {
    
    Set<List<Integer>> result = new HashSet<>();
    int nums[];
    
    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        
        this.nums = nums;
        traverse(new ArrayList(), 0);
        
        List<List<Integer>> list = new ArrayList();
        for(List<Integer> i : result)
            list.add(i);
        return list;
        
        
    }
    
    void traverse(List<Integer> list, int index){
        if(index >= nums.length){
            result.add(new ArrayList(list));
            return;
        }
        list.add(nums[index]);
        traverse(list, index+1);
        list.remove(Integer.valueOf(nums[index]));
        traverse(list, index+1);
        
    }
    
}
